File: web/src/lib/money.ts
/**
 * Exact money helpers. Amounts are integers in minor units; the API accepts
 * and returns decimal strings. Mirrors backend config/masroof.php.
 */
export const CURRENCY_EXPONENTS: Record<string, number> = {
  SAR: 2,
  AED: 2,
  KWD: 3,
  BHD: 3,
  OMR: 3,
  QAR: 2,
  EGP: 2,
  JOD: 3,
  USD: 2,
  EUR: 2,
  GBP: 2,
  TRY: 2,
  MAD: 2,
  TND: 3,
  DZD: 2,
  IQD: 3,
  LBP: 2,
  PKR: 2,
  INR: 2,
  IDR: 2,
  MYR: 2,
  JPY: 0,
};

export const CURRENCIES = Object.keys(CURRENCY_EXPONENTS);

const ARABIC_SYMBOLS: Record<string, string> = {
  SAR: "ر.س",
  AED: "د.إ",
  KWD: "د.ك",
  BHD: "د.ب",
  OMR: "ر.ع",
  QAR: "ر.ق",
  EGP: "ج.م",
  JOD: "د.أ",
  IQD: "د.ع",
  MAD: "د.م",
  TND: "د.ت",
  DZD: "د.ج",
  LBP: "ل.ل",
};

export function exponentOf(currency: string): number {
  const exponent = CURRENCY_EXPONENTS[currency];
  if (exponent === undefined) throw new Error(`Unsupported currency ${currency}`);
  return exponent;
}

export function normalizeDigits(input: string): string {
  return input.replace(/[٠-٩]/g, (d) => String(d.charCodeAt(0) - 0x0660)).replace(/[۰-۹]/g, (d) => String(d.charCodeAt(0) - 0x06f0));
}

/** Parses user input into minor units, or null when invalid for the currency. */
export function parseAmount(input: string, currency: string, { allowNegative = false } = {}): number | null {
  const exponent = exponentOf(currency);
  let value = normalizeDigits(input)
    .replace(/[\s,٬]/g, "")
    .replace("٫", ".");
  const negative = allowNegative && value.startsWith("-");
  if (negative) value = value.slice(1);
  const pattern = exponent === 0 ? /^\d{1,15}$/ : new RegExp(`^\\d{1,15}(\\.\\d{1,${exponent}})?$`);
  if (!pattern.test(value)) return null;
  const [whole, fraction = ""] = value.split(".");
  const minor = Number(whole + fraction.padEnd(exponent, "0"));
  if (!Number.isSafeInteger(minor)) return null;
  return negative ? -minor : minor;
}

export function toDecimal(minor: number, currency: string): string {
  const exponent = exponentOf(currency);
  const digits = String(Math.abs(minor)).padStart(exponent + 1, "0");
  const value = exponent === 0 ? digits : `${digits.slice(0, -exponent)}.${digits.slice(-exponent)}`;
  return minor < 0 ? `-${value}` : value;
}

export function currencySymbol(currency: string, locale: string): string {
  return locale.startsWith("ar") ? (ARABIC_SYMBOLS[currency] ?? currency) : currency;
}

/** Grouped amount with Western digits and bidi isolation for RTL text. */
export function formatMoney(
  minor: number,
  currency: string,
  locale: string,
  { signed = false, symbol = true }: { signed?: boolean; symbol?: boolean } = {},
): string {
  const [whole, fraction] = toDecimal(Math.abs(minor), currency).split(".");
  const grouped = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  const number = fraction ? `${grouped}.${fraction}` : grouped;
  const sign = minor < 0 ? "-" : signed && minor > 0 ? "+" : "";
  if (!symbol) return `${sign}${number}`;
  const text = `⁦${sign}${number}⁩`;
  return locale.startsWith("ar") ? `${text} ${currencySymbol(currency, locale)}` : `${currencySymbol(currency, locale)} ${text}`;
}

File: mobile/lib/core/money/money.dart
import 'package:intl/intl.dart';

/// Currency metadata and exact minor-unit conversions (no floating point).
/// Mirrors backend config/masroof.php.
abstract final class Money {
  static const exponents = <String, int>{
    'SAR': 2, 'AED': 2, 'KWD': 3, 'BHD': 3, 'OMR': 3, 'QAR': 2, //
    'EGP': 2, 'JOD': 3, 'USD': 2, 'EUR': 2, 'GBP': 2, 'TRY': 2, //
    'MAD': 2, 'TND': 3, 'DZD': 2, 'IQD': 3, 'LBP': 2, 'PKR': 2, //
    'INR': 2, 'IDR': 2, 'MYR': 2, 'JPY': 0,
  };

  static const _arabicSymbols = <String, String>{
    'SAR': 'ر.س', 'AED': 'د.إ', 'KWD': 'د.ك', 'BHD': 'د.ب', 'OMR': 'ر.ع', //
    'QAR': 'ر.ق', 'EGP': 'ج.م', 'JOD': 'د.أ', 'IQD': 'د.ع', 'MAD': 'د.م', //
    'TND': 'د.ت', 'DZD': 'د.ج', 'LBP': 'ل.ل',
  };

  static List<String> get currencies => exponents.keys.toList();

  static int exponent(String currency) {
    final value = exponents[currency];
    if (value == null) throw ArgumentError.value(currency, 'currency');
    return value;
  }

  /// Parses user input ("1,250.5", "١٢٫٥", "12.50") into minor units.
  /// Returns null when the input is not a valid amount for the currency.
  static int? tryParse(String input, String currency) {
    final exp = exponent(currency);
    final normalized = normalizeDigits(input).replaceAll(RegExp(r'[\s,٬]'), '').replaceAll('٫', '.');
    final pattern = exp == 0 ? RegExp(r'^\d{1,15}$') : RegExp('^\\d{1,15}(\\.\\d{1,$exp})?\$');
    if (!pattern.hasMatch(normalized)) return null;

    final parts = normalized.split('.');
    final fraction = parts.length > 1 ? parts[1] : '';
    return int.parse(parts[0] + fraction.padRight(exp, '0'));
  }

  /// Exact decimal string accepted by the API, e.g. 150050 SAR -> "1500.50".
  static String toDecimal(int minor, String currency) {
    final exp = exponent(currency);
    final negative = minor < 0;
    final digits = minor.abs().toString().padLeft(exp + 1, '0');
    final value = exp == 0
        ? digits
        : '${digits.substring(0, digits.length - exp)}.${digits.substring(digits.length - exp)}';
    return negative ? '-$value' : value;
  }

  /// Human readable amount with grouping. Western digits are used in both
  /// languages for unambiguous financial figures.
  static String format(
    int minor,
    String currency, {
    required String locale,
    bool withSymbol = true,
    bool signed = false,
  }) {
    final exp = exponent(currency);
    final pattern = NumberFormat.decimalPatternDigits(locale: 'en', decimalDigits: exp);
    final whole = minor.abs() ~/ _pow10(exp);
    final fraction = minor.abs() % _pow10(exp);
    var text = pattern.format(whole);
    if (exp > 0) {
      text = '${text.split('.').first}.${fraction.toString().padLeft(exp, '0')}';
    }

    final sign = minor < 0 ? '-' : (signed && minor > 0 ? '+' : '');
    if (!withSymbol) return '$sign$text';

    final symbol = symbolFor(currency, locale);
    // Unicode isolates keep sign, number and symbol ordered inside RTL text.
    return locale.startsWith('ar') ? '\u2066$sign$text\u2069 $symbol' : '$symbol \u2066$sign$text\u2069';
  }

  static String symbolFor(String currency, String locale) {
    if (locale.startsWith('ar')) return _arabicSymbols[currency] ?? currency;
    return currency;
  }

  /// Converts Arabic-Indic and Eastern Arabic-Indic digits to ASCII.
  static String normalizeDigits(String input) {
    final buffer = StringBuffer();
    for (final rune in input.runes) {
      if (rune >= 0x0660 && rune <= 0x0669) {
        buffer.writeCharCode(0x30 + rune - 0x0660);
      } else if (rune >= 0x06F0 && rune <= 0x06F9) {
        buffer.writeCharCode(0x30 + rune - 0x06F0);
      } else {
        buffer.writeCharCode(rune);
      }
    }
    return buffer.toString();
  }

  static int _pow10(int exp) {
    var result = 1;
    for (var i = 0; i < exp; i++) {
      result *= 10;
    }
    return result;
  }
}
